File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {        
        int N = maps.length;
        int M = maps[0].length;
        
        //bfs
        
        int[] dr = {-1,1,0,0};
        int[] dc = {0,0,-1,1};
        
        // Set<String> visited = new HashSet<>(); // visited set
        boolean[][] visited = new boolean[N][M];
        Queue<int[]> queue = new LinkedList<>(); // queue
        
        queue.add(new int[]{0,0,1});
        visited[0][0] = true;
        
        while(!queue.isEmpty()) {
            int[] now = queue.poll();
            int r = now[0];
            int c = now[1];
            int dist = now[2];
            
            if(r == (N-1) && c == (M-1)) {
                return dist;
            }
            
            
            for(int d=0; d<4; d++) {
                int nr = dr[d] + r;
                int nc = dc[d] + c;
                
                if(0<=nr && nr<N &&
                  0<=nc && nc<M &&
                  maps[nr][nc] == 1 &&
                  !visited[nr][nc]) {
                    
                    visited[nr][nc] = true;
                    queue.add(new int[]{nr, nc, dist+1});

                }
                
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    
    dr = [-1,1,0,0]
    dc = [0,0,-1,1]
    
    N = len(maps)
    M = len(maps[0])
    
    visited = [[0]*M for _ in range(N)]
    
    dq = deque()
    dq.append((0,0,1))   
    visited[0][0] = 1
    
    while dq:
        r, c, cnt = dq.popleft()
        
        if r==N-1 and c==M-1 :
            return cnt
            break
        
        for d in range(4):
            nr = dr[d] + r
            nc = dc[d] + c
            
            if 0<=nr<N and 0<=nc<M and visited[nr][nc] == 0 and maps[nr][nc] == 1 :
                dq.append((nr,nc,cnt+1))
                visited[nr][nc] = 1
                
    else: return -1       
    return cnt